File: ai-worker/LlmClient.cs
using System.Diagnostics;
using System.Text;
using System.Text.Json;

namespace AiWorker;

public sealed record LlmResult(
    string Content,
    int? PromptTokens,
    int? CompletionTokens,
    string? FinishReason,
    long ElapsedMs);

public sealed class LlmClient
{
    private readonly HttpClient http;
    private readonly LlmOptions options;

    public LlmClient(HttpClient http, LlmOptions options)
    {
        this.http = http;
        this.options = options;
        http.Timeout = TimeSpan.FromSeconds(options.TimeoutSeconds);
        if (!string.IsNullOrEmpty(options.ApiKey))
            http.DefaultRequestHeaders.Add("Authorization", $"Bearer {options.ApiKey}");
    }

    public Task<LlmResult> CompleteAsync(string system, string user, CancellationToken ct)
    {
        return CompleteAsync(system, user, VerdictSchema.ResponseFormat, 0, ct);
    }

    public async Task<LlmResult> CompleteAsync(string system, string user, string responseFormat, int seedOffset, CancellationToken ct)
    {
        var request = new Dictionary<string, object?>
        {
            ["model"] = options.Model,
            ["messages"] = new object[]
            {
                new Dictionary<string, object?> { ["role"] = "system", ["content"] = system },
                new Dictionary<string, object?> { ["role"] = "user", ["content"] = user },
            },
            ["temperature"] = options.Temperature,
            ["seed"] = options.Seed + seedOffset,
            ["stream"] = false,
            ["max_tokens"] = options.MaxCompletionTokens,
            ["response_format"] = JsonDocument.Parse(responseFormat).RootElement.Clone(),
        };
        var body = JsonSerializer.Serialize(request);
        var watch = Stopwatch.StartNew();

        HttpResponseMessage response;
        try
        {
            response = await http.PostAsync("chat/completions",
                new StringContent(body, Encoding.UTF8, "application/json"), ct);
        }
        catch (TaskCanceledException) when (!ct.IsCancellationRequested)
        {
            throw new LlmCallException($"model generation timed out after {options.TimeoutSeconds}s");
        }
        catch (OperationCanceledException)
        {
            throw;
        }
        catch (HttpRequestException ex)
        {
            throw new LlmUnavailableException($"model endpoint unreachable: {ex.Message}");
        }
        catch (Exception ex)
        {
            throw new LlmCallException($"model call failed before a response: {ex.Message}");
        }

        using (response)
        {
            string raw;
            try
            {
                raw = await response.Content.ReadAsStringAsync(ct);
            }
            catch (OperationCanceledException) when (ct.IsCancellationRequested)
            {
                throw;
            }
            catch (Exception ex)
            {
                throw new LlmCallException($"model response read failed: {ex.Message}");
            }

            watch.Stop();

            if (!response.IsSuccessStatusCode)
                throw new LlmCallException(
                    $"model endpoint returned {(int)response.StatusCode}: {Snippet(raw)}");

            return Extract(raw) with { ElapsedMs = watch.ElapsedMilliseconds };
        }
    }

    internal static LlmResult Extract(string raw)
    {
        try
        {
            using var doc = JsonDocument.Parse(raw);
            var root = doc.RootElement;
            if (root.ValueKind == JsonValueKind.Object &&
                root.TryGetProperty("choices", out var choices) &&
                choices.ValueKind == JsonValueKind.Array &&
                choices.GetArrayLength() > 0)
            {
                var choice = choices[0];
                if (choice.ValueKind == JsonValueKind.Object &&
                    choice.TryGetProperty("message", out var header) &&
                    header.ValueKind == JsonValueKind.Object &&
                    header.TryGetProperty("content", out var content) &&
                    content.ValueKind == JsonValueKind.String)
                {
                    return new LlmResult(
                        content.GetString()!,
                        TokenCount(root, "prompt_tokens"),
                        TokenCount(root, "completion_tokens"),
                        choice.TryGetProperty("finish_reason", out var finish) && finish.ValueKind == JsonValueKind.String
                            ? finish.GetString()
                            : null,
                        0);
                }
            }
        }
        catch (JsonException)
        {
        }
        throw new ModelOutputException("chat response missing choices[0].message.content", raw);
    }

    private static int? TokenCount(JsonElement root, string property)
    {
        if (!root.TryGetProperty("usage", out var usage) || usage.ValueKind != JsonValueKind.Object)
            return null;
        if (!usage.TryGetProperty(property, out var value) || value.ValueKind != JsonValueKind.Number)
            return null;
        return value.TryGetInt32(out var tokens) ? tokens : null;
    }

    private static string Snippet(string text)
    {
        return text.Length <= 200 ? text : $"{text[..200]}...";
    }
}

File: assistant-extension/controllers/llm-client.js
console.log("ASSISTANT", "llm-client");

class LlmClient {

    static REQUEST_TIMEOUT = 180000;
    static MAX_TOKENS = 8192;

    constructor(url, model) {
        this.url = url;
        this.model = model;
    }

    static async Create() {
        return new LlmClient(await StorageHandler.LlamaUrlAsync(), await StorageHandler.LlamaModelAsync());
    }

    async Chat(messages, tools, options) {
        const controller = new AbortController();
        const timer = setTimeout(function () { controller.abort(); }, LlmClient.REQUEST_TIMEOUT);

        try {
            const response = await fetch(this.url + "v1/chat/completions", {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                signal: controller.signal,
                body: JSON.stringify({
                    model: this.model,
                    messages: messages,
                    tools: tools,
                    temperature: 0.2,
                    stream: false,
                    max_tokens: (options && options.max_tokens) || LlmClient.MAX_TOKENS,
                })
            });

            const body = await response.text();

            if (!response.ok) {
                console.error("ASSISTANT", "LlmClient", "http", response.status, body.slice(0, 200));
                return { error: "llm-http-" + response.status };
            }

            return LlmClient.Parse(body);
        } catch (e) {
            const error = e && e.name === 'AbortError' ? "llm-timeout" : "llm-unreachable";
            console.error("ASSISTANT", "LlmClient", error, e);
            return { error: error };
        } finally {
            clearTimeout(timer);
        }
    }

    static Parse(raw) {
        let parsed;
        try {
            parsed = JSON.parse(raw);
        } catch (e) {
            return { error: "llm-invalid-json" };
        }

        const message = parsed && parsed.choices && parsed.choices[0] && parsed.choices[0].message;
        if (!message) return { error: "llm-no-choice" };

        return {
            content: message.content ?? "",
            tool_calls: (message.tool_calls || []).map(function (call) {
                let args = {};
                try {
                    args = JSON.parse(call.function?.arguments || "{}");
                } catch (e) {
                    args = { __parse_error: true };
                }
                return { id: call.id, name: call.function?.name, args: args };
            })
        };
    }
}
